### PR TITLE
Refactor sitemap CSV import around a canonical resolved plan (#767)

### PR DESCRIPTION
**Pull request by flesher**
Reviewable diff: +2135/-1594 across 6 files (excludes generated, test, and story files).

## Summary

Site-map CSV import lets an operator upload a spreadsheet of sites, buildings, racks, and miners and have Proto Fleet reconcile the fleet's topology to match it. Previously each stage of that flow — the dry-run preview, the validation errors, the "did anything change since you previewed" token, and the final database writes — worked out *from scratch* what the CSV meant. Those four independent re-derivations drifted apart, which is where most of the reported edge-case bugs came from.

This PR makes the import resolve the CSV into **one canonical plan** first, and then has preview, validation, the commit token, and apply all read from that single plan. Same feature, same CSV format from the user's point of view; the internals now have a single source of truth instead of four.

## How it works

The import runs in one direction: raw CSV → resolved plan → everything else.

1. **Parse.** The uploaded bytes are split into the four sections (SITE, BUILDING, RACK, MINER), each a list of rows.

2. **Load a snapshot.** The server reads what currently exists in the database: all sites, buildings, and racks for the org, plus the miners the uploader is allowed to see. Miners the uploader *can't* see but which still occupy a rack slot are kept in a separate "hidden members" list so they're never silently trampled or deleted.

3. **Resolve references.** Every parent link in the CSV is a single cell, read like this:
   - blank → unassigned
   - a bare number (e.g. `building` = `12`) → the existing building with id 12
   - `NAME:x` (e.g. `building` = `NAME:Warehouse B`) → a building being *created in this same import* whose name is "Warehouse B"

   This replaces the old scheme where each link was two columns (a name *and* an id) that could disagree. Resolving also fills in implied ancestors (a rack that names a building inherits that building's site) and records the matched building's real id internally, so two different buildings that happen to share the same (site, name) stay distinct.

4. **Build the plan.** Each row becomes a resolved node (site / building / rack / miner) tagged with an action — **none**, **create**, or **update** — decided by comparing the canonical CSV row against the live row field by field. Alongside the nodes, a "topology view" describes the desired end state (CSV merged over the current snapshot) that the validators check against.

5. **Validate, count, tokenize — all off the plan.** Grid-position collisions, rack and building capacity, slot conflicts and bounds, and rename permissions all read the resolved nodes and the shared topology view (and fold in those hidden members). The preview change-counts come from the node actions. The commit token is a hash of the resolved plan plus the live snapshot fingerprint.

6. **Apply.** On commit, the token is re-checked (if the CSV or the live database changed since the dry-run, the token won't match and the commit is refused). Then a single database transaction walks the resolved nodes and performs each create/update, and — because "remove omitted rows" is now the only removal mode — deletes anything the CSV left out.

Example of the token guard: you dry-run and get a token. A teammate moves a rack in the UI before you hit commit. Your token no longer matches the fresh snapshot, so the commit is rejected and you're asked to dry-run again, rather than applying a stale plan.

```mermaid
flowchart TD
    CSV["Uploaded CSV"] --> Parse["Parse into SITE/BUILDING/RACK/MINER rows"]
    DB["Live database"] --> Snap["Load snapshot (org topology + visible miners + hidden members)"]
    Parse --> Resolve["Resolve references (single-cell: blank / id / NAME:x)"]
    Snap --> Resolve
    Resolve --> Plan["Resolved plan: nodes tagged none/create/update + topology view"]
    Plan --> Preview["Preview change counts"]
    Plan --> Validate["Validation (grid, capacity, slots, renames)"]
    Plan --> Token["Commit token = hash(plan + snapshot)"]
    Plan --> Apply["Apply: one transaction, walk nodes, delete omitted"]
    Apply --> DB
```

```mermaid
sequenceDiagram
    actor User
    participant Import as Import service
    participant DB
    User->>Import: Upload CSV (dry_run)
    Import->>DB: Load snapshot
    Import->>Import: Resolve plan, validate, count
    Import-->>User: Preview + commit token
    User->>Import: Commit (with token)
    Import->>DB: Reload snapshot
    Import->>Import: Recompute token
    alt token matches
        Import->>DB: Apply plan in one transaction
        Import-->>User: Success + changes
    else snapshot drifted
        Import-->>User: Rejected — dry-run again
    end
```

## Areas of the code involved

| Area / package / file | What changed | Why it matters for review |
| --- | --- | --- |
| `server/internal/domain/sitemap/resolved.go` | **New.** The canonical plan: resolved node types, `resolveReferences`, `resolvePlan`, the topology view, action classification, and all validators that now read the plan. | This is the heart of the PR — the single source of truth everything else consumes. |
| `server/internal/domain/sitemap/service.go` | Rewritten around the plan: export is id-authoritative, preview/token/apply read the plan, and the old per-stage re-derivation helpers are deleted (net ~-900 lines). | Confirm preview, token, and apply all derive from the plan and that the deleted helpers have no remaining callers. |
| `server/internal/domain/buildings/models/models.go` | **New helpers** `GridCapacity` and `RackCapacityExceeded` — the building rack-capacity rule extracted so both the buildings domain and sitemap import share one implementation. | A previously duplicated invariant; check the two callers agree on semantics. |
| `server/internal/domain/buildings/service.go` | Uses the extracted capacity helper instead of its own inline arithmetic. | Behavior-preserving; verify the net-membership count passed in is unchanged. |
| `client/.../SiteMapCsvImportModal.tsx` | Removes the "leave omitted rows in place" radio option; the omission prompt now just confirms removal. | Matches the server: remove-omitted is the only committable behavior. |
| `docs/plans/2026-07-22-767-sitemap-canonical-plan-tdd.md` | **New** design/plan document. | Optional context for the refactor's phasing and out-of-scope items. |

## Key technical decisions & trade-offs

- **Single reference cell (id or `NAME:`) instead of a name+id column pair.** The old pair could contradict itself and needed reconciliation logic; one cell removes that whole class of ambiguity. Trade-off: existing entities must be referenced by id, and same-import creations by `NAME:` — bare names are a validation error.
- **Id-precise building identity via an internal companion cell.** Two buildings can legitimately share a (site, name) pair, so the resolved building id is carried alongside the canonical name to keep them distinct in grid/capacity/move checks. Alternative — keying purely on (site, name) — silently merged them.
- **Remove-omitted is the only removal mode.** "Leave in place" was dropped as a committable option (server and client). Simpler contract; the dead `OMISSION_MODE_LEAVE_IN_PLACE` proto enum value is left for a follow-up because proto regen is unreliable in this worktree.
- **Commit token hashes the resolved plan, not the raw rows + preview summary.** A token can no longer certify a topology that apply would resolve differently. Trade-off: token values change (they're opaque to clients, so this is safe).
- **Extracted the rack-capacity invariant into the buildings domain** rather than keeping sitemap's private copy, so the rule has one home. Kept the signature capacity-agnostic (`aisles, racksPerAisle, resultingCount`) so a future interactive-reparent fix (#778) can reuse it.

## Testing & validation

- `go build ./...`, `go vet`, and `gofmt -l` clean; the full `internal/domain/sitemap` package test suite passes, as do the new `buildings/models` capacity tests.
- Behavior is asserted against the resolved plan so the previously-reported edge cases can't silently regress, including: renamed/moved buildings staying visible to the rack grid/capacity validators; id-disambiguation of duplicate (site, name) buildings; hidden rack members counted in slot/capacity/conflict checks; and remove-omitted never unassigning a hidden miner.
- Commit-token behavior is covered both ways: the token changes when the live snapshot drifts, and when a CSV edit changes the resolved plan even against an identical snapshot.
- **Not covered:** deletion of the now-dead `OMISSION_MODE_LEAVE_IN_PLACE` proto enum value (deferred — needs proto regen outside a worktree); no new end-to-end/integration test beyond the existing service-level suite.

### CODE REVIEW

File: server/internal/domain/sitemap/service.go [hunk 1]
```diff
@@ -334,15 +345,15 @@ func (s *Service) ImportSiteMapCsv(ctx context.Context, orgID int64, req *pb.Imp
 			Warnings:       plan.warnings,
 		}, nil
 	}
-	if hasOmissions(plan.omissions) && req.GetOmissionMode() == pb.OmissionMode_OMISSION_MODE_UNSPECIFIED {
+	if hasOmissions(plan.omissions) && req.GetOmissionMode() != pb.OmissionMode_OMISSION_MODE_REMOVE_OMITTED {
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/sitemap/service.go:348]: 🤖 Working as intended. Remove-omitted is now the only committable omission mode — a locked product decision (commit f3c2a7c20). The leave-in-place path was removed behaviorally; the preview surfaces the omitted counts and asks for confirmation before removal. The dead `OMISSION_MODE_LEAVE_IN_PLACE` proto enum value is slated for deletion in a follow-up (proto regen is unreliable in this worktree, so the enum cleanup is tracked separately).

File: server/internal/domain/sitemap/resolved.go [hunk 2]
```diff
@@ -0,0 +1,1076 @@
...
+
+	pb "github.com/block/proto-fleet/server/generated/grpc/sitemap/v1"
+	buildingmodels "github.com/block/proto-fleet/server/internal/domain/buildings/models"
+	sitemodels "github.com/block/proto-fleet/server/internal/domain/sites/models"
+)
+
+// resolved.go builds the canonical resolved plan for a sitemap import: parsed
+// rows and the live snapshot are collapsed into one graph of typed nodes keyed
+// by stable ID where one exists, with parents linked as pointers, so preview,
+// validation, the commit token, and apply all consume a single representation.
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/sitemap/resolved.go]: 🤖 Fixed in 18421d4f0. Removed the dead parent-pointer fields (`resolvedBuilding.site`, `resolvedRack.building/site`, `resolvedMiner.rack/building/site`) and rewrote the header comment to describe the actual model: parents are carried as name/id references, not pointers.

File: server/internal/domain/sitemap/resolved.go [hunk 3]
```diff
@@ -0,0 +1,1076 @@
...
+	return errs
+}
+
+// validateKnownMiners rejects rows whose device identifier is not a live miner.
+func validateKnownMiners(miners []*resolvedMiner) []*pb.ImportValidationError {
+	var errs []*pb.ImportValidationError
+	for _, m := range miners {
+		if m.deviceID != "" && m.existing == nil {
+			errs = append(errs, csvErr(m.rowNum, "MINER", "unknown miner device_identifier"))
+		}
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/sitemap/resolved.go:387]: 🤖 Fixed in 18421d4f0. The error now names the offending device: `unknown miner device_identifier %q`.

File: server/internal/domain/sitemap/service_test.go [hunk 4]
```diff
@@ -40,18 +40,16 @@ func TestBuildSiteMapExportZipIncludesCSVAndAgentGuide(t *testing.T) {
...
 	guideText := files[siteMapExportGuideTXTPath]
 	for _, want := range []string{
 		"Edit proto-fleet-site-map/site-map.csv",
-		"If rack is set, the rack determines the miner's building and site.",
-		"ID columns identify existing records or disambiguate references.",
+		"If rack is set, the rack determines the miner's building and site; leave building and site blank.",
 		"Leave omitted rows in place keeps missing rows unchanged.",
 		"Remove omitted rows soft-deletes omitted sites, buildings, and racks, and unassigns omitted miners.",
-		"Prefer name references unless an ID is needed to disambiguate.",
 	} {
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/sitemap/service_test.go:52]: 🤖 Fixed in 18421d4f0. Both the exported guide text and its test assertion were updated to the remove-omitted-only wording (dropped the "Leave omitted rows in place" line).

**Comment by flesher** [comment]
🤖 Triaged both Codex Security Review findings:

**Finding 1 — leave-in-place omission mode (MEDIUM):** Working as intended. Remove-omitted is now the only committable omission mode (locked product decision, commit f3c2a7c20). The commit path rejects any other mode, and the preview surfaces the omitted counts and requires explicit confirmation before removal. The dead `OMISSION_MODE_LEAVE_IN_PLACE` proto enum value is slated for deletion in a follow-up (proto regen is unreliable in this worktree, so the enum cleanup is tracked separately).

**Finding 2 — silent parent-reference overwrite (MEDIUM):** Fixed in 18421d4f0. This is intended precedence, not a silent bug: for a single row, a rack reference wins over a building reference, which wins over a site reference. The finding pointed at dead id-vs-name mismatch validation branches that could never fire under the single-reference-cell model. Those dead branches are now deleted and the rack-wins precedence is documented inline in `validatePlacementConsistency`.

Independent security review: LOW risk, no new vulnerabilities. Import stays inside a single `RunInTx`; the commit token is an optimistic-concurrency drift check recomputed and compared server-side. One pre-existing consideration noted — during import the plan reflects org-wide topology (sites/buildings/racks load org-wide, only miners are viewer-scoped), consistent with the accepted #520 decision; not changed here.

File: server/internal/domain/sitemap/service.go [hunk 5]
```diff
@@ -2680,42 +2598,25 @@ func validateRemoveOmittedReferences(parsed *parsedCSV, mode pb.OmissionMode) []
...
-		}
-		if row[fieldBuilding] == "" {
-			continue
-		}
-		if id, ok := idFromCell(row[fieldBuildingID]); ok && presentBuildingIDs[id] {
-			continue
-		}
-		if row[fieldSite] != "" {
+		if row[fieldBuilding] != "" {
 			if !presentBuildings[row[fieldSite]+"\x00"+row[fieldBuilding]] {
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/sitemap/service.go:2632]: 🤖 Valid — fixed in afaf8b5e9. This bug lives entirely inside REMOVE_OMITTED, so the "only committable mode" enforcement did not cover it.

`validateRemoveOmittedReferences` only checked that some `(site, name)` BUILDING row remained. Since two unassigned buildings can share a `(site, name)` pair via distinct NULL `site_id`s, a rack/miner referencing the omitted twin by numeric id passed validation (the retained same-name twin satisfied the name check), got placed under the omitted id by `applyRacks`/`applyMiners`, then orphaned when `applyOmittedRows` deleted it in the same transaction.

Fix: the check now keys on the resolved id preserved in `refBuildingIDCell` against the retained BUILDING rows own ids, not the canonicalized name. Bare-id references get the id check; NAME:/name references (which leave the cell nil) keep the name check. Regression test `TestValidateRemoveOmittedReferencesRejectsOmittedBuildingTwin` constructs the two-same-name-unassigned-buildings case and asserts the omitted-twin reference is rejected while the retained-twin reference is accepted.

File: server/internal/domain/sitemap/resolved.go [hunk 6]
```diff
@@ -0,0 +1,1029 @@
...
+import (
+	"fmt"
+	"strconv"
+
+	pb "github.com/block/proto-fleet/server/generated/grpc/sitemap/v1"
+	buildingmodels "github.com/block/proto-fleet/server/internal/domain/buildings/models"
+	sitemodels "github.com/block/proto-fleet/server/internal/domain/sites/models"
+)
+
+// resolved.go builds the canonical resolved plan for a sitemap import: parsed
```
Comment (ankitgoswami) [inline server/internal/domain/sitemap/resolved.go:12]: nit: something like import_plan would have been clearer

File: server/internal/domain/sitemap/resolved.go [hunk 7]
```diff
@@ -0,0 +1,1029 @@
...
+	siteRows := parsed.sections["SITE"]
+	buildingRows := parsed.sections["BUILDING"]
+	rackRows := parsed.sections["RACK"]
+	tv := &topologyView{
+		sites:                  desiredSiteSet(siteRows, target.sites),
+		buildingKeys:           rowSetFromDesiredBuildings(buildingRows, target.buildings),
+		buildingsByKey:         desiredBuildingMap(buildingRows, target.buildings),
+		buildingsByLayoutID:    desiredBuildingLayoutIDMap(buildingRows, target.buildings),
+		buildingsByCapacityKey: desiredBuildingCapacityMap(buildingRows, target.buildings),
+		racksByLabel:           desiredRackMap(rackRows, target.racks, buildingRows, target.buildings),
```
Comment (ankitgoswami) [inline server/internal/domain/sitemap/resolved.go]: I think we can simplify and DRY this up a bit, the helpers seem to be doing similar operations. Maybe call desiredBuildingList once, then iterate over the desiredBuildings and create the different maps (buildingsByKey, buildingsByLayoutID, buildingsByCapacityKey). 
Can also pass buildingsByKey and buildingsByID to desiredRackMap so it can use them to resolve rack parents.

This way there will be fewer places to edit if we need to change the csv in the future

**Review by ankitgoswami — APPROVED** [review]

File: server/internal/domain/sitemap/resolved.go (hunk 6 above)
Comment (flesher) [reply server/internal/domain/sitemap/resolved.go:12]: 🤖 Good instinct, but `importPlan` is already taken — it is a distinct struct in `service.go` (the preview/errors/warnings response container). `resolvedPlan`/`resolved.go` deliberately names the *canonical resolved* layer to keep the two apart. Renaming would either collide or force renaming both across the whole package for a cosmetic gain, so leaving as-is.

File: server/internal/domain/sitemap/resolved.go (hunk 7 above)
Comment (flesher) [reply server/internal/domain/sitemap/resolved.go]: 🤖 Done in cd0725323. `resolveTopologyView` now calls `desiredBuildingList` once and derives the (site,name), id, and capacity-key indexes in a single pass, passing the shared (site,name)+id maps into `desiredRackMap`. Removed the five parallel row+snapshot walks (`desiredBuildingMap`, `desiredBuildingLayoutIDMap`, `desiredBuildingCapacityMap`, `desiredBuildingsBySiteName`, `desiredBuildingsByID`) — net −92 lines, so there is one place to edit when the CSV changes. Sitemap suite (golden corpus + twin-building cases) unchanged and green.

File: server/internal/domain/sitemap/resolved.go [hunk 8]
```diff
@@ -0,0 +1,1047 @@
...
+	for _, site := range snap.sites {
+		row := rowMap(siteHeaders, siteRawRows([]sitemodels.Site{site})[0])
+		siteExisting[siteIdentity(site)] = row
+		siteExisting["name:"+site.Name] = row
+	}
+	buildingExisting := map[string]map[string]string{}
+	for _, building := range snap.buildings {
+		row := rowMap(buildingHeaders, buildingRawRows([]buildingmodels.Building{building})[0])
+		buildingExisting[buildingIdentity(building)] = row
+		buildingExisting["name:"+building.SiteLabel+"\x00"+building.Name] = row
```
[the remark this answers is not in the document]
Comment (flesher) [reply server/internal/domain/sitemap/resolved.go:852]: 🤖 Valid observation, and accurate — confirmed against the code and the DB. A blank-id BUILDING row sharing `(site, name)` with an existing **unassigned** building is name-aliased in `classifyTopologyActions`, so it classifies as update/no-op instead of create. No server error (the classify and apply paths share the same `(site,name)` key, so no crash/`NotFound`); the effect is silent — the intended second unassigned building is dropped or the existing one is mutated in place, and `NAME:` refs follow it.

This is really the sharp edge of a broader inconsistency: sites and racks are org-wide name-unique, but unassigned buildings are intentionally *not* name-unique (`uk_building_site_name` is `WHERE site_id IS NOT NULL`), so "name == identity" only half-holds. Rather than patch the one symptom inside this refactor PR, tracking it as a follow-up to unify the duplicate-name model (DB + import identity/classification) across all three entities: https://github.com/block/proto-fleet/issues/814 (priority: low). Resolving here.
